_gsocblogs/2022/blog_ROOT_JunZhang.md

```diff
@@ -12,24 +13,80 @@ intro: |
+When ROOT sees `edm`, it will immediately pull up all modules containing it, which is totally unnecessary since we only care about `X`. So my project this year is aiming to eliminate the unnecessary lookup for identifiers like a namespace to reduce memory pressure.
 
-## Future Direction
+## The Work I have done
 
-The next step is to verify how large an impact our work has made especially in a bigger project (for example CMSSW). I have applied to a CMSSW external account and will do deeper profiling after I'm granted CMSSW resources. Another issue we need to solve is that the current approach only works for top-level namespaces and we need to do more to make it work for all-level namespaces. When all of these are done, we'll also try to upstream our work to LLVM main repo.
+At the very beginning of the project, I tried to eliminate all namespaces identifiers lookups as it's the most straightforward idea. But we soon ran into some troubles. The problem is that we lost the ability to recognize identifiers inside those nested namespaces, which is a very common case:
 
-## Summary
+```c++
+ error: no member named 'MakeTrivialDataFrame' in namespace 'ROOT::RDF'
+   auto d_s = ROOT::RDF::MakeTrivialDataFrame(nEvents);
+              ~~~~~~~~~~~^
+```
 
-This is my first time participating in the open source community this deeply. It has been a wonderful experience for me and I have learned a lot from it. I would like to give my most sincere thanks to those who helped me here.
+This failure looks hard to solve so I was thinking: Why not achieve our goals step by step and only take top-level namespaces into consideration? So I did. The test results look good but when it comes to the performance improvement, Ugh... the results doesn't look very satisfing. The reason is simple, as there're not too many top-level namspaces, so we can't get significant memory decrease.
 
-If you get this far, thanks for reading! See you next month!
+We seem to touch some dark corners in ROOT. During that period, I explored many directions but none seems to work. My mentors and I had some deep discussions, and finally, we came to a new approach: If we cannot ignore those namespaces, why not create a new module that contains all namespaces forward declarations? Because the manually generated module just consists of forward declarations, it's cheap to preload, and ROOT can have all essential knowledge about namespaces with it. So here's the [PR](https://github.com/root-project/root/pull/10910). The performance improvement also looks very promising:
```

Comment:
Dark corners of ROOT but also the semantics of C++ does not allow us to do much even outside of ROOT.